File: stats/outlierScore.R
library(extraDistr)
library(stringr)

args = commandArgs(trailingOnly=TRUE)
fitFile=args[1]
countsFile=args[2]
outname=args[3]

psiStats=read.table(fitFile,sep="\t",header = TRUE,stringsAsFactors = FALSE)

data=read.csv(countsFile)
iso1_idx=which(endsWith(colnames(data),"iso1"))
iso2_idx=which(endsWith(colnames(data),"iso2"))
sample_names=str_replace(colnames(data)[iso1_idx],"_iso1","")

sampleCount=length(sample_names)
eventCount=nrow(data)

print(paste('sampleCount:',sampleCount))
print(paste('eventCount:',eventCount))

iso1=sapply(data[,iso1_idx],as.numeric)
iso2=sapply(data[,iso2_idx],as.numeric)
info_idx=seq(1,iso1_idx[1]-1)

idx=match(data$event_jid,psiStats$event_jid)
print(idx[which(!is.na(idx))])
alpha=psiStats[idx,"alpha"]
beta=psiStats[idx,"beta"]
scores=matrix(NA,nrow = length(alpha),ncol = sampleCount)
colnames(scores)<-sample_names
idx1=which(alpha>=beta)
idx2=which(alpha<beta)
if(length(idx1)>0){
  scores[idx1,]=t(sapply(idx1, function(x) log(pmax(pbbinom(iso1[x,],iso1[x,]+iso2[x,],alpha[x],beta[x],log.p=FALSE),.Machine$double.xmin))))
}
if(length(idx2)>0){
  scores[idx2,]=t(sapply(idx2,function(x) -log(pmax(pbbinom(iso2[x,],iso1[x,]+iso2[x,],beta[x],alpha[x],log.p=FALSE),.Machine$double.xmin))))
}

write.table(cbind(data[,info_idx],scores),outname,row.names=FALSE,append = FALSE,sep="\t")

File: stats/outlierScore.py
import sys, pandas
from scipy.stats import betabinom
import numpy as np
from itertools import compress
from operator import add

fit_file = sys.argv[1]
counts_file =sys.argv[2]
outname = sys.argv[3]+".bisbeeOutlier.csv"

psi_stats = pandas.read_csv(fit_file)

data = pandas.read_csv(counts_file)
iso1_idx = [col for col in data.columns if 'iso1' in col]
iso2_idx = [col for col in data.columns if 'iso2' in col]
sample_names= [x.replace("_iso1","") for x in iso1_idx]

sample_count=len(sample_names)
event_count = data.shape[0]

print("sampleCount: "+str(sample_count))
print("eventCount: "+str(event_count))

iso1=data[iso1_idx]
iso2=data[iso2_idx]
info_idx = [col for col in data.columns if 'iso1' not in col and 'iso2' not in col]

idx= psi_stats.reset_index().set_index('event_jid').loc[data["event_jid"], 'index'].values
alpha = psi_stats.alpha[idx] #
beta = psi_stats.beta[idx]
scores = pandas.DataFrame(index=range(0,len(alpha)), columns = sample_names)

idx1= (alpha>=beta)
idx1= list(compress(range(len(idx1)),idx1))
idx2= (alpha<beta)
idx2= list(compress(range(len(idx2)),idx2))

if(len(idx1) >0):
	scores.iloc[idx1]=list(map(lambda x: np.log(np.maximum(betabinom.cdf(iso1.iloc[x],list(map(add, iso1.iloc[x],iso2.iloc[x])),alpha[x],beta[x]),sys.float_info.min)),idx1))
if(len(idx2)>0):
	scores.iloc[idx2]=list(map(lambda x: -np.log(np.maximum(betabinom.cdf(iso2.iloc[x],list(map(add, iso1.iloc[x],iso2.iloc[x])),beta[x],alpha[x]),sys.float_info.min)),idx2))
iso1.columns =iso2.columns
scores[iso1+iso2 ==0]=float("nan")

dfout=pandas.concat([data[info_idx],scores], axis=1)
dfout.to_csv(outname)
